src/ast/ast.cc

```diff
@@ -42,7 +65,29 @@ namespace ast
       auto find = std::find(parent->nodes.begin(), parent->nodes.end(), ast);
       assert(find != parent->nodes.end());
       parent->nodes.erase(find);
+      ast->parent.reset();
+    }
+  }
+
+  void rename(Ast& ast, const char* name)
+  {
+    Ast next;
+
+    if (ast->is_token)
+    {
+      next = token(ast, name, ast->token);
     }
+    else
+    {
+      next = node(ast, name);
+
+      for (auto& node : ast->nodes)
+        push_back(next, node);
```

Comment:
This should probably be `std::move(node)` to avoid unnecessary atomic incref/decref.

Reply:
It works now. Optimisations can be made in a future PR.